Add tests of RandomDataset

frreiss opened the pull request:
This PR adds tests for the RandomDataset kernel, following the general approach from @feihugis's latest updates to the `DatasetOpsTestBaseV2` test harness class.

I also split `random_dataset_op.cc` into separate `.cc` and `.h` files so that the test cases can instantiate the kernel.

frreiss committed:
Refactor RandomDataset into .h and .cc files

Refactor RandomDataset into .h and .cc files

Cleanup after refactoring

frreiss committed:
Add initial version of test case

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Refactor to new format

frreiss committed:
Fix typo in human-readable node name

rachellim reviewed (changes requested):
Can you use the `ITERATOR_GET_NEXT_TEST_P`, `DATASET_NODE_NAME_TEST_P`, etc macros that @feihugis added?

feihugis commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,249 @@
+  TF_ASSERT_OK(Initialize(&test_case.dataset_params));
+
+  // Can't use DatasetOpsTestBase::CheckIteratorGetNext because the kernel
+  // under test produces unbounded input.
```
@rachellim @frreiss As `CheckIteratorGetNext()` does not work here, I think `ITERATOR_GET_NEXT_TEST_P` will not work for this test either.

frreiss replied:
Yes, the current version of `ITERATOR_GET_NEXT_TEST_P` is not appropriate for infinite data sets. I can use the other macros though.

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Refactor test cases to new format

frreiss commented:
> Can you use the `ITERATOR_GET_NEXT_TEST_P`, `DATASET_NODE_NAME_TEST_P`, etc macros that @feihugis added?

Sure, can do. Commit 0a70de6 modifies the test cases that have corresponding macros to use the macros (with the exception of the test for `GetNext`, as noted in the conversation with @feihugis above).

mellanox-github commented:
Can one of the admins verify this patch?

rachellim reviewed (dismissed):
Thanks!

rthadur commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,269 @@
+#include "tensorflow/core/kernels/data/experimental/random_dataset_op.h"
+
+#include "tensorflow/core/kernels/data/dataset_test_base.h"
+#include "tensorflow/core/lib/random/philox_random.h"
```
error: module //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/random/philox_random.h'

frreiss replied:
This dependency was only reflected as a transitive dependency in the build file. I've added a direct dependency on the new `//tensorflow/core/lib/random:philox` file set, which includes `philox_random.h` in commit 1d547b2.

rthadur commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,269 @@
+
+#include "tensorflow/core/kernels/data/dataset_test_base.h"
+#include "tensorflow/core/lib/random/philox_random.h"
+#include "tensorflow/core/lib/random/random.h"
```
error: module //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/random/random.h'

frreiss replied:
I was able to remove this include after pulling the latest changes from master.

rthadur commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,269 @@
+#include "tensorflow/core/kernels/data/dataset_test_base.h"
+#include "tensorflow/core/lib/random/philox_random.h"
+#include "tensorflow/core/lib/random/random.h"
+#include "tensorflow/core/lib/random/random_distributions.h"
```
error: module //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/random/random_distributions.h'

frreiss replied:
The dependency on the `//tensorflow/core/lib/random:philox` that I added in in commit 1d547b2 includes `random_distributions.h`.

rthadur reviewed (changes requested):
failed internall test with below errors , can you please check

frreiss commented:
Looking into the build file issues.

frreiss committed:
Add direct dependencies to test target

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Update build rules to use new random:philox target

frreiss commented:
Build files updated, then updated a second time to reflect changes in master.

jsimsa commented:
```
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:55:65: error: only virtual member functions can be marked 'override'
  Status MakeInputs(gtl::InlinedVector<TensorValue, 4>* inputs) override {
                                                                ^~~~~~~~~
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:65:36: error: unknown template name 'DatasetOpsTestBaseV2'
class RandomDatasetOpTest : public DatasetOpsTestBaseV2<RandomDatasetParams> {
                                   ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:67:58: error: only virtual member functions can be marked 'override'
  Status Initialize(RandomDatasetParams* dataset_params) override {
                                                         ^~~~~~~~~
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:105:68: error: only virtual member functions can be marked 'override'
                             std::unique_ptr<OpKernel>* op_kernel) override {
                                                                   ^~~~~~~~~
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:69:39: error: use of undeclared identifier 'thread_num_'
    TF_RETURN_IF_ERROR(InitThreadPool(thread_num_));
                                      ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:70:55: error: use of undeclared identifier 'cpu_num_'
    TF_RETURN_IF_ERROR(InitFunctionLibraryRuntime({}, cpu_num_));
                                                      ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:79:62: error: use of undeclared identifier 'dataset_kernel_'
    TF_RETURN_IF_ERROR(MakeDatasetOpKernel(*dataset_params, &dataset_kernel_));
                                                             ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:84:30: error: use of undeclared identifier 'dataset_kernel_'
        CreateDatasetContext(dataset_kernel_.get(), &inputs, &dataset_ctx_));
                             ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:84:63: error: use of undeclared identifier 'dataset_ctx_'
        CreateDatasetContext(dataset_kernel_.get(), &inputs, &dataset_ctx_));
                                                              ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:89:23: error: use of undeclared identifier 'dataset_kernel_'
        CreateDataset(dataset_kernel_.get(), dataset_ctx_.get(), &dataset_));
                      ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:89:46: error: use of undeclared identifier 'dataset_ctx_'
        CreateDataset(dataset_kernel_.get(), dataset_ctx_.get(), &dataset_));
                                             ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:89:67: error: use of undeclared identifier 'dataset_'
        CreateDataset(dataset_kernel_.get(), dataset_ctx_.get(), &dataset_));
                                                                  ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:94:31: error: use of undeclared identifier 'dataset_ctx_'
        CreateIteratorContext(dataset_ctx_.get(), &iterator_ctx_));
                              ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:94:52: error: use of undeclared identifier 'iterator_ctx_'
        CreateIteratorContext(dataset_ctx_.get(), &iterator_ctx_));
                                                   ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:95:24: error: use of undeclared identifier 'dataset_'
    TF_RETURN_IF_ERROR(dataset_->MakeIterator(iterator_ctx_.get(),
                       ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:95:47: error: use of undeclared identifier 'iterator_ctx_'
    TF_RETURN_IF_ERROR(dataset_->MakeIterator(iterator_ctx_.get(),
                                              ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:96:65: error: use of undeclared identifier 'iterator_'; did you mean 'IteratorResource::State::iterator'?
                                              kIteratorPrefix, &iterator_));
                                                                ^~~~~~~~~
                                                                IteratorResource::State::iterator
./third_party/tensorflow/core/lib/core/errors.h:73:43: note: expanded from macro 'TF_RETURN_IF_ERROR'
    const ::tensorflow::Status _status = (__VA_ARGS__);  \
                                          ^
./third_party/tensorflow/core/kernels/data/iterator_ops.h:82:35: note: 'IteratorResource::State::iterator' declared here
    std::unique_ptr<IteratorBase> iterator;
                                  ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:111:57: error: reference to non-static member function must be called; did you mean to call it with no arguments?
        {{RandomDatasetOp::kOutputTypes, dataset_params.output_dtypes},
                                         ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
                                                                     ()
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:112:58: error: reference to non-static member function must be called; did you mean to call it with no arguments?
         {RandomDatasetOp::kOutputShapes, dataset_params.output_shapes}});
                                          ~~~~~~~~~~~~~~~^~~~~~~~~~~~~
                                                                      ()
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Update test case to work with new version of test harness

frreiss commented:
I've merged the latest updates from `master` and added some changes that should resolve the build issues.

feihugis commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,237 @@
+    return Status::OK();
+  }
+
+  virtual Status GetInputs(
```
@frreiss After merging #32750 , `GetInputs()` is changed to `std::vector<Tensor> GetInputTensors() const override`, which only returns the input tensors (no need to return the dataset tensor). [Here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc#L40-L45) is an example. For this case, it will be:
```C++
std::vector<Tensor> GetInputTensors() const override {
  return {seed_, seed2_};
}

frreiss replied:
Fixed this in commit bb2a02e

feihugis commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,237 @@
+        seed_(CreateTensor<int64>(TensorShape({}), {seed})),
+        seed2_(CreateTensor<int64>(TensorShape({}), {seed2})) {}
+
+  virtual Status GetInputPlaceholder(
```
After this change (https://github.com/tensorflow/tensorflow/commit/d9f05892a72db8a9dca1682642cbad33486b9f81), `GetInputPlaceholder(...)` is renamed to `GetInputNames(...)`.

frreiss replied:
Made this change in commit bb2a02e

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Update code towork with latest version of test harness

rachellim commented on tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:
```diff
@@ -0,0 +1,232 @@
+
+  virtual Status GetInputNames(
+      std::vector<string>* input_placeholder) const override {
+    *input_placeholder = {RandomDatasetOp::kSeed, RandomDatasetOp::kSeed2};
```
s/placeholder/names

frreiss replied:
Good catch, fixing that now.

frreiss replied:
Fixed in 708bc86

frreiss committed:
Rename argument to match underlying API

rthadur commented:
@frreiss can you please check build failures ?

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Fix problem with fix for build problem

frreiss commented:
I've put in another layer of changes in the BUILD files that this PR modifies. The build is now working again on my test machines. Hopefully it works on all the different CI targets now.

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Update test case to match latest version of test harness

rthadur commented:
`third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:15:10: error: module //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/random/philox_random.h'

build_cleaner //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test
#include "third_party/tensorflow/core/lib/random/philox_random.h"
         ^
third_party/tensorflow/core/kernels/data/experimental/random_dataset_op_test.cc:16:10: error: module //third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/random/random_distributions.h'
//third_party/tensorflow/core/kernels/data/experimental:random_dataset_op_test
#include "third_party/tensorflow/core/lib/random/random_distributions.h"`

@frreiss can you please check above error ?

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-random-test

frreiss committed:
Hopefully fix build_cleaner issue

frreiss committed:
Update test to reflect latest test harness

frreiss commented:
I have added some dependencies to the `random_dataset_op_test` target that hopefully will clear up the error described above.

I also updated one of the test cases to conform to the latest version of the test harness.

rthadur commented:
@rachellim can you please review new changes ?

rthadur commented:
This has merged internally , waiting for auto-merge to happen, thank you.
